projects/chaos/src/sequencer.c

```diff
@@ -0,0 +1,231 @@
+    { .id = CHAOS_EVENT_CLOSE_RELAY, .data = RELAY_ID_SOLAR_MASTER_REAR },
+    { .id = CHAOS_EVENT_NO_OP, .data = EMPTY_DATA },
+    { .id = CHAOS_EVENT_RELAY_CLOSED, .data = RELAY_ID_SOLAR_MASTER_REAR },
+    { .id = CHAOS_EVENT_CLOSE_RELAY, .data = RELAY_ID_SOLAR_MASTER_FRONT },
+    { .id = CHAOS_EVENT_NO_OP, .data = EMPTY_DATA },
+    { .id = CHAOS_EVENT_RELAY_CLOSED, .data = RELAY_ID_SOLAR_MASTER_FRONT },
+    { .id = CHAOS_EVENT_CLOSE_RELAY, .data = RELAY_ID_MAIN_POWER },
+    { .id = CHAOS_EVENT_NO_OP, .data = EMPTY_DATA },
+    { .id = CHAOS_EVENT_RELAY_CLOSED, .data = RELAY_ID_MAIN_POWER },
+    { .id = CHAOS_EVENT_MONITOR_ENABLE, .data = POWER_PATH_SOURCE_ID_DCDC },
+    { .id = CHAOS_EVENT_SEQUENCE_DRIVE_DONE, .data = EMPTY_DATA },
+  };
+  s_event_idx = 0;
+  s_sizeof_events = SIZEOF_ARRAY(drive_events);
+  s_events = drive_events;
+}
+
+void sequencer_init(void) {
+  s_retries = 0;
+  fsm_state_init(sequencer_state_emergency, prv_sequencer_state_emergency);
+  fsm_state_init(sequencer_state_idle, prv_sequencer_state_idle);
+  fsm_state_init(sequencer_state_charge, prv_sequencer_state_charge);
+  fsm_state_init(sequencer_state_drive, prv_sequencer_state_drive);
+  fsm_init(&s_sequencer_fsm, "SequenceFSM", &sequencer_state_idle, NULL);
+  // Invalid event_idx since vehicle starts IDLE.
+  s_event_idx = UINT16_MAX;
+}
+
+StatusCode sequencer_publish_next_event(const Event *previous_event) {
+  if (previous_event->id <= NUM_CHAOS_EVENTS_CAN &&
+      !(previous_event->id == CHAOS_EVENT_SEQUENCE_EMERGENCY ||
+        previous_event->id == CHAOS_EVENT_SEQUENCE_EMERGENCY_DONE)) {
+    // Not in the range of handled events. This isn't necessarily invalid but won't be handled so
+    // return unknown.
+    return STATUS_CODE_UNKNOWN;
+  } else if (previous_event->id == CHAOS_EVENT_NO_OP) {
+    return STATUS_CODE_OK;
+  }
+
+  // Try to transition if the event is a sequence transition.
+  if (previous_event->id == CHAOS_EVENT_SEQUENCE_IDLE ||
+      previous_event->id == CHAOS_EVENT_SEQUENCE_CHARGE ||
+      previous_event->id == CHAOS_EVENT_SEQUENCE_DRIVE ||
+      previous_event->id == CHAOS_EVENT_SEQUENCE_EMERGENCY ||
+      previous_event->id == CHAOS_EVENT_SEQUENCE_RESET) {
+    if (!fsm_process_event(&s_sequencer_fsm, previous_event)) {
+      LOG_DEBUG("Err failed transition: %u %u\n", previous_event->id, previous_event->data);
+      return status_code(STATUS_CODE_INVALID_ARGS);
+    }
+  }
+
+  // Stop if there are no more events.
+  if (s_event_idx >= s_sizeof_events) {
+    s_retries = 0;
+    return status_code(STATUS_CODE_RESOURCE_EXHAUSTED);
+  }
+
+  // If the s_event_idx == 0 we have no need to check the previous event.
+  if (s_event_idx == 0) {
+    event_raise(s_events[s_event_idx].id, s_events[s_event_idx].data);
+    s_event_idx++;
+    return STATUS_CODE_OK;
+  }
+
+  // Check that the previous event was expected or restart the sequence otherwise.
+  // NOTE: this implicitly means any events in the range
```

Comment:
Should probably make sure this is in the header - I was a little confused at first but I understand what the idea behind this is now.

Reply:
Done.